### PR TITLE
XCFA-refactor 2023 [draft]

### PR DESCRIPTION
**Pull request by leventeBajczi**
This PR merges the `xcfa-refactor` branch.

Main features of the PR:

- [x] Adds back support for the [llvm](https://llvm.org/) backend for XCFAs
- [x] Refactors the analysis methods for programs over weakly ordered architectures
- [x] Adds the refactored runtime monitor (by @AdamZsofi )
- [x] Adds support for the [ARG web debugger](https://ftsrg-edu.github.io/student-sisak-argviz/)
- [x] Adds serialization and deserialization to all Exprs and Stmts, as well as ARGs and related data structures
- [x] Adds a new type of solver (`graph-solver`) to Theta
- [x] Adds a new frontend, `litmus-cli` for litmus test inputs
- [x] Adds lots of new tests and resources for better quality assurance throughout Theta
- [x] Rewrites the XCFA type family to use Kotlin and data classes for easier use and maintainability
- [x] Rewrites the XCFA pass system to be more resilient to reconfiguration    
- [x] Expand the documentation of various subprojects of Theta

### CODE REVIEW

**Comment by leventeBajczi** [comment]
@csanadtelbisz: Sonar fails, because some tests ultimately fail.
Relevant log:
``` 
Caused by:
2023-09-10T13:13:40.7201330Z         java.lang.UnsupportedOperationException: Solver Z3 not supported
2023-09-10T13:13:40.7201918Z             at hu.bme.mit.theta.solver.SolverManager.resolveSolverFactory(SolverManager.java:40)
2023-09-10T13:13:40.7202556Z             at hu.bme.mit.theta.xcfa.passes.LoopUnrollPass.<clinit>(LoopUnrollPass.kt:46)
2023-09-10T13:13:40.7202939Z             ... 5 more
```
Can you take a look if this has something to do with the moved registerSolverFactories method in XcfaCli? I think I saw some modifications there in your pr (#202). Let me know if I'm mistaken.

**Comment by csanadtelbisz** [comment]
@leventeBajczi The solver issue arose due to the fact that `LoopUnrollPass` invokes a solver (unprecedented in xcfa passes before). Tests did not register solvers. I added a dependency on Z3 and hardcoded the usage of Z3 in `LoopUnrollPass`.
Also fixed bugs reported by sonarcloud.

**Comment by szdan97** [comment]
A projection should be given in XcfaAnalysis.getXcfaAbstractor in order to avoid performance degradation compared to the original xcfa analysis, as the lack of non-trivial projection leads to a lot of unnecessary covering checks

**Comment by csanadtelbisz** [comment]
The WebDebuggerLogger is always enabled and it is responsible for a considerable portion of runtime. See the following lines: https://github.com/ftsrg/theta/blob/0c8bc7c48fc89bce04ac1bdd331ad8af7dfc55ff/subprojects/common/analysis/src/main/java/hu/bme/mit/theta/analysis/algorithm/cegar/CegarChecker.java#L98-L101
As it serves only debugging purposes, it should be disabled by default.
@AdamZsofi git blame shows that you are the author of this change.

The following profiler output confirms the problem. The WebDebuggerLogger takes ~1/3 (more than 10 seconds in this case) of all CPU time.
image

**Comment by csanadtelbisz** [comment]
Wrong result is given for the unreach-call property on the following tasks in the eca-rers2012 category of sv-benchmarks:

Config 1: `--strategy DIRECT --maxenum 1 --loglevel RESULT --refinement SEQ_ITP --lbe LBE_SEQ --search ERR --prunestrategy FULL --predsplit ATOMS --cex-monitor DISABLE --domain EXPL`
Problem17_label11.yml

Config 2: `--strategy DIRECT --maxenum 1 --loglevel RESULT --refinement SEQ_ITP --lbe LBE_SEQ --search ERR --prunestrategy FULL --predsplit ATOMS --cex-monitor DISABLE --domain PRED_CART`
Problem10_label30.yml, Problem11_label40.yml, Problem15_label49.yml

File: subprojects/frontends/llvm/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,97 @@
+/*
```
Comment (as3810t) [inline subprojects/frontends/llvm/build.gradle.kts:1]: What is the relation of this module to `libtheta-llvm.so` in the `lib`?

File: subprojects/xcfa/litmus-cli/src/main/java/hu/bme/mit/theta/xcfa/litmus/cli/LitmusCli.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+                final XCFA xcfa = LitmusInterpreter.getXcfa(litmus);
+                logger.write(Logger.Level.MAINSTEP, "Litmus test parsed successfully\n");
+
+                if (printxcfa) {
+                    System.out.println("digraph G{");
+                    //TODO visualize
+                    System.out.println("}");
+                }
+
+//            final List<Integer> processIds = listToRange(processes, -1, -1);
```
Comment (as3810t) [inline subprojects/xcfa/litmus-cli/src/main/java/hu/bme/mit/theta/xcfa/litmus/cli/LitmusCli.java:134]: Major block of code commented out

File: subprojects/common/analysis/src/main/java/hu/bme/mit/theta/analysis/algorithm/mcm/analysis/FiniteStateChecker.kt [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import hu.bme.mit.theta.analysis.expr.StmtAction
+import hu.bme.mit.theta.common.Tuple
+import hu.bme.mit.theta.common.Tuple1
+import hu.bme.mit.theta.common.Tuple2
+import hu.bme.mit.theta.graphsolver.compilers.GraphPatternCompiler
+import hu.bme.mit.theta.graphsolver.patterns.constraints.GraphConstraint
+import hu.bme.mit.theta.graphsolver.solvers.GraphSolver
+import java.util.*
+
+class FiniteStateChecker<S : ExplState, A : StmtAction, T>(
```
Comment (as3810t) [inline subprojects/common/analysis/src/main/java/hu/bme/mit/theta/analysis/algorithm/mcm/analysis/FiniteStateChecker.kt:36]: Please add documentation (or links to papers) that describe what this algorithm can be used for.

File: subprojects/common/analysis/src/main/java/hu/bme/mit/theta/analysis/algorithm/mcm/analysis/PartialSolver.kt [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+package hu.bme.mit.theta.analysis.algorithm.mcm.analysis
+
+import hu.bme.mit.theta.common.Tuple
+import hu.bme.mit.theta.graphsolver.ThreeVL
+import hu.bme.mit.theta.graphsolver.collectSubRelations
+import hu.bme.mit.theta.graphsolver.compilers.GraphPatternCompiler
+import hu.bme.mit.theta.graphsolver.patterns.constraints.GraphConstraint
+import hu.bme.mit.theta.graphsolver.solvers.GraphSolver
+
+class PartialSolver<T>(
```
Comment (as3810t) [inline subprojects/common/analysis/src/main/java/hu/bme/mit/theta/analysis/algorithm/mcm/analysis/PartialSolver.kt:26]: Please add documentation (or links to papers) that describe what this algorithm can be used for.

File: subprojects/xcfa/cat/src/main/java/hu/bme/mit/theta/cat/solver/DatalogSolver.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package hu.bme.mit.theta.cat.solver;
+
+public class DatalogSolver {
```
Comment (as3810t) [inline subprojects/xcfa/cat/src/main/java/hu/bme/mit/theta/cat/solver/DatalogSolver.java:19]: Is this class really needed?

**Review by as3810t — CHANGES_REQUESTED** [review]
All in all, this is an immense PR. I would like to ask for clarification on the following major things:

- With the (re)introduction of the LLVM frontend, Theta had its first cpp module. What is the relation between the `frontends/llvm` submodule, the `libtheta-llvm.so` library, and the `theta-c-frontend` repository?
  - If I interpret it correctly, `theta-c-frontend` was migrated into  `frontends/llvm`, but without its project history, and `libtheta-llvm.so` is the built version of `frontends/llvm`
  - Can we / should we migrate the history of `theta-c-frontend`?
  - Should we ship with a prebuilt version of `libtheta-llvm.so`?
- The XCFA `ProcedurePassManager` seems to enable no passes for CHC's. Is it intentional, or a regression error?
- Other minor things in comments.

**Comment by leventeBajczi** [comment]
Thanks for the review! I'll get around to incorporating its findings hopefully this weekend.

**Review by csanadtelbisz — CHANGES_REQUESTED** [review]
@AdamZsofi  WebDebuggerLogger still has an overhead, see the following lines: https://github.com/ftsrg/theta/blob/fdc40cf95ba46bb8a94a791a785f031db2010a2a/subprojects/common/analysis/src/main/java/hu/bme/mit/theta/analysis/algorithm/cegar/CegarChecker.java#L94-L95

Time taken by these lines:
image

**Comment by AdamZsofi** [comment]
> @AdamZsofi WebDebuggerLogger still has an overhead, see the following lines:
> 
> https://github.com/ftsrg/theta/blob/fdc40cf95ba46bb8a94a791a785f031db2010a2a/subprojects/common/analysis/src/main/java/hu/bme/mit/theta/analysis/algorithm/cegar/CegarChecker.java#L94-L95
> 
> Time taken by these lines: image

Thanks for telling me. Fixed, there should be no overhead now.

**Review by csanadtelbisz — DISMISSED** [review]

**Review by szdan97 — APPROVED** [review]

**Pull request merged**